Bump Maps, NN, `common`, and mapbox-java dependency versions to `10.1.0`, `80.0.2`, `20.1.0`, and 6.2.0 respectively

Guardiola31337 opened the pull request:
### Description
<!--
Include issue references (e.g., fixes [#issue](link))
Include necessary implementation details (e.g. I opted to use this algorithm because ... and test it in this way ...).
-->
Bumps Maps, NN and `common` dependency versions to `10.1.0`, `80.0.2` and `20.1.0` respectively

Guardiola31337 commented:
I updated OP with Maps `v10.1.0`, NN `v80.0.2` and `common` `v20.1.0`. This is ready for another round of 👀 

cc @LukasPaczos @RingerJK @korshaknn

RingerJK commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/LocationEx.kt:
```diff
@@ -48,11 +51,32 @@ internal fun Location.toFixLocation(): FixLocation {
 private fun isCurrentSdkVersionEqualOrGreaterThan(sdkCode: Int): Boolean =
     Build.VERSION.SDK_INT >= sdkCode
+
+internal fun HashMap<String, Value>.toBundle(): Bundle? {
```
👍

RingerJK reviewed (approved):
LGTM

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/RouteAlternativesController.kt:
```diff
@@ -89,6 +90,10 @@ internal class RouteAlternativesController constructor(
             return emptyList()
         }
+
+        override fun onError(message: String) {
+            logE(TAG, Message("Error: $message"))
+        }
```
Are there any cases where we should call the observer with empty routes here (to cleanup the state after error)?

Guardiola31337 replied:
Discussing this with @LukasPaczos we have some questions @mapbox/navnative:

- are there situations where we should clean everything up?
- what type of errors can it report? just failed requests or something more?

SiarheiFedartsou replied:
cc @DenisPryt

DenisPryt replied:
We report message from RouterError now (i.e. we report only about RouterError) 
If this error occurs you don't need to clean up the state (At least now). If we get this error it means that we've been trying to request new routes but something went wrong.

DenisPryt replied:
onRouteAlternativesChanged will be invoked after alternatives changed regardless of the errors that occur.
If we pass alternatives after error, onRouteAlternativesChanged will be invoked. Or if alternatives will changed for any reasons (refreshImmediately or timeout) -- onRouteAlternativesChanged will be invoked (only if alternatives will be changed)

Guardiola31337 replied:
Awesome! Thanks for confirming @DenisPryt 🙇‍♂️

Guardiola31337 commented:
A couple of issues found by `instrumentation-tests` 💪 

```
11-10 15:03:32.516: W/Bundle(28059): Key mockLocation expected String but value was a java.lang.Boolean.  The default value <null> was returned.
11-10 15:03:32.517: W/Bundle(28059): Attempt to cast generated internal exception:
11-10 15:03:32.517: W/Bundle(28059): java.lang.ClassCastException: java.lang.Boolean cannot be cast to java.lang.String
11-10 15:03:32.517: W/Bundle(28059): 	at android.os.BaseBundle.getString(BaseBundle.java:923)
11-10 15:03:32.517: W/Bundle(28059): 	at com.mapbox.navigation.core.navigator.LocationEx.toMap(LocationEx.kt:79)
11-10 15:03:32.517: W/Bundle(28059): 	at com.mapbox.navigation.core.navigator.LocationEx.toFixLocation(LocationEx.kt:55)
11-10 15:03:32.517: W/Bundle(28059): 	at com.mapbox.navigation.core.trip.session.MapboxTripSession$updateRawLocation$2.invokeSuspend(MapboxTripSession.kt:241)
11-10 15:03:32.517: W/Bundle(28059): 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
11-10 15:03:32.517: W/Bundle(28059): 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
11-10 15:03:32.517: W/Bundle(28059): 	at android.os.Handler.handleCallback(Handler.java:739)
11-10 15:03:32.517: W/Bundle(28059): 	at android.os.Handler.dispatchMessage(Handler.java:95)
11-10 15:03:32.517: W/Bundle(28059): 	at androidx.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:148)
11-10 15:03:32.517: W/Bundle(28059): 	at androidx.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:519)
11-10 15:03:32.517: W/Bundle(28059): 	at androidx.test.espresso.base.UiControllerImpl.loopMainThreadUntilIdle(UiControllerImpl.java:441)
11-10 15:03:32.517: W/Bundle(28059): 	at androidx.test.espresso.Espresso$1.run(Espresso.java:277)
11-10 15:03:32.517: W/Bundle(28059): 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:423)
11-10 15:03:32.517: W/Bundle(28059): 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
11-10 15:03:32.517: W/Bundle(28059): 	at android.os.Handler.handleCallback(Handler.java:739)
11-10 15:03:32.517: W/Bundle(28059): 	at android.os.Handler.dispatchMessage(Handler.java:95)
11-10 15:03:32.517: W/Bundle(28059): 	at android.os.Looper.loop(Looper.java:148)
11-10 15:03:32.517: W/Bundle(28059): 	at android.app.ActivityThread.main(ActivityThread.java:5417)
11-10 15:03:32.517: W/Bundle(28059): 	at java.lang.reflect.Method.invoke(Native Method)
11-10 15:03:32.517: W/Bundle(28059): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
11-10 15:03:32.517: W/Bundle(28059): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
11-10 15:03:32.519: W/Bundle(28059): Key mockLocation expected String but value was a java.lang.Boolean.  The default value <null> was returned.
11-10 15:03:32.520: W/Bundle(28059): Attempt to cast generated internal exception:
11-10 15:03:32.520: W/Bundle(28059): java.lang.ClassCastException: java.lang.Boolean cannot be cast to java.lang.String
11-10 15:03:32.520: W/Bundle(28059): 	at android.os.BaseBundle.getString(BaseBundle.java:923)
11-10 15:03:32.520: W/Bundle(28059): 	at com.mapbox.navigation.core.navigator.LocationEx.toMap(LocationEx.kt:79)
11-10 15:03:32.520: W/Bundle(28059): 	at com.mapbox.navigation.core.navigator.LocationEx.toFixLocation(LocationEx.kt:55)
11-10 15:03:32.520: W/Bundle(28059): 	at com.mapbox.navigation.core.trip.session.MapboxTripSession$updateRawLocation$2.invokeSuspend(MapboxTripSession.kt:241)
11-10 15:03:32.520: W/Bundle(28059): 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
11-10 15:03:32.520: W/Bundle(28059): 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
11-10 15:03:32.520: W/Bundle(28059): 	at android.os.Handler.handleCallback(Handler.java:739)
11-10 15:03:32.520: W/Bundle(28059): 	at android.os.Handler.dispatchMessage(Handler.java:95)
11-10 15:03:32.520: W/Bundle(28059): 	at androidx.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:148)
11-10 15:03:32.520: W/Bundle(28059): 	at androidx.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:519)
11-10 15:03:32.520: W/Bundle(28059): 	at androidx.test.espresso.base.UiControllerImpl.loopMainThreadUntilIdle(UiControllerImpl.java:441)
11-10 15:03:32.520: W/Bundle(28059): 	at androidx.test.espresso.Espresso$1.run(Espresso.java:277)
11-10 15:03:32.520: W/Bundle(28059): 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:423)
11-10 15:03:32.520: W/Bundle(28059): 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
11-10 15:03:32.520: W/Bundle(28059): 	at android.os.Handler.handleCallback(Handler.java:739)
11-10 15:03:32.520: W/Bundle(28059): 	at android.os.Handler.dispatchMessage(Handler.java:95)
11-10 15:03:32.520: W/Bundle(28059): 	at android.os.Looper.loop(Looper.java:148)
11-10 15:03:32.520: W/Bundle(28059): 	at android.app.ActivityThread.main(ActivityThread.java:5417)
11-10 15:03:32.520: W/Bundle(28059): 	at java.lang.reflect.Method.invoke(Native Method)
11-10 15:03:32.520: W/Bundle(28059): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
11-10 15:03:32.520: W/Bundle(28059): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
11-10 15:03:32.532: D/MbxNavigationTelemetry(28059): session state is ActiveGuidance(sessionId=327864ae-6949-409f-bced-e4d99f11a77d)
11-10 15:03:32.532: D/MbxNavigationTelemetry(28059): trackFreeDrive STOP
11-10 15:03:32.533: D/MbxNavigationTelemetry(28059): createFreeDriveEvent STOP
11-10 15:03:32.537: D/WifiService(873): acquireWifiLockLocked: WifiLock{NlpWifiLock type=2 binder=android.os.BinderProxy@f76191e}
11-10 15:03:32.538: D/MbxNavigationTelemetry(28059): populateFreeDriveEvent
11-10 15:03:32.539: D/MbxNavigationTelemetry(28059): class com.mapbox.navigation.core.telemetry.events.NavigationFreeDriveEvent event sent
11-10 15:03:32.540: D/MbxNavigationTelemetry(28059): onRoutesChanged. Number of routes = 2; reason = ROUTES_UPDATE_REASON_NEW
11-10 15:03:32.540: D/MbxNavigationTelemetry(28059): handle a new route
11-10 15:03:32.540: D/MbxNavigationTelemetry(28059): resetOriginalRoute
11-10 15:03:32.540: D/MbxNavigationTelemetry(28059): resetRouteProgress
11-10 15:03:32.544: D/wifi(873): Initialized common fields 10000, 16, 100, 10
11-10 15:03:32.544: D/wifi(873): bucket[0] = 2:3:10000:2
11-10 15:03:32.615: I/nav-native(28059): Waypoints were not passed to setRoute. Trying to restore it from route itself.
11-10 15:03:32.615: W/nav-native(28059): Cannot find via_waypoints in passed route. On-board waypoints matching will be used.
11-10 15:03:32.615: W/nav-native(28059): Cannot find via_waypoints in passed route. On-board waypoints matching will be used.
11-10 15:03:32.615: I/nav-native(28059): Successful route response. Metadata: 
11-10 15:03:32.617: I/NearbyDiscovery(23206): Show notifications: 0 total, no changes since last shown, no-op. [CONTEXT service_id=49 ]
11-10 15:03:32.629: D/MbxNavigationTelemetry(28059): populateNavigationEvent
11-10 15:03:32.630: D/MbxNavigationTelemetry(28059): class com.mapbox.navigation.core.telemetry.events.NavigationDepartEvent event sent
11-10 15:03:32.640: I/art(28059): Note: end time exceeds epoch: 
11-10 15:03:32.687: W/System.err(28059): java.lang.NullPointerException: Attempt to invoke virtual method 'com.mapbox.api.directions.v5.models.DirectionsRoute$Builder com.mapbox.api.directions.v5.models.DirectionsRoute.toBuilder()' on a null object reference
11-10 15:03:32.693: W/System.err(28059): 	at com.mapbox.api.directions.v5.models.DirectionsResponse.fromJson(DirectionsResponse.java:169)
11-10 15:03:32.693: W/System.err(28059): 	at com.mapbox.navigation.base.internal.utils.DirectionsResponseUtilsKt$parseDirectionsResponse$2.invokeSuspend(DirectionsResponseUtils.kt:29)
11-10 15:03:32.693: W/System.err(28059): 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
11-10 15:03:32.693: W/System.err(28059): 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
11-10 15:03:32.693: W/System.err(28059): 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
11-10 15:03:32.693: W/System.err(28059): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
11-10 15:03:32.693: W/System.err(28059): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
11-10 15:03:32.693: W/System.err(28059): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
11-10 15:03:32.693: E/nav-native(28059): RouteAlternativesControllerImpl unhandled observer unknown exception in onRouteAlternativesChanged method
11-10 15:03:32.704: E/DEBUG(28059): alternatives failed history path=/data/user/0/com.mapbox.navigation.instrumentation_tests/files/mbx_nav/history/2021-11-10T23-03-32Z_a936487f-a766-4791-9388-79d5e41dc120.pbf.gz
11-10 15:03:32.704: W/LocationUpdateReceiver(28059): LocationEngineResult == null
11-10 15:03:32.705: I/MbxHistoryRecorder(28059): Initial size is 4096 bytes
11-10 15:03:32.709: D/MbxNavigationTelemetry(28059): MapboxNavigation onDestroy
11-10 15:03:32.712: D/MbxNavigationTelemetry(28059): session state is com.mapbox.navigation.core.trip.session.NavigationSessionState$Idle@5cb37ed
11-10 15:03:32.712: D/MbxNavigationTelemetry(28059): handleSessionCanceled
11-10 15:03:32.713: D/MbxNavigationTelemetry(28059): flush buffer. Pending events = 0
11-10 15:03:32.724: D/MbxNavigationTelemetry(28059): populateNavigationEvent
11-10 15:03:32.725: D/MbxNavigationTelemetry(28059): class com.mapbox.navigation.core.telemetry.events.NavigationCancelEvent event sent
11-10 15:03:32.725: D/MbxNavigationTelemetry(28059): resetOriginalRoute
11-10 15:03:32.725: D/MbxNavigationTelemetry(28059): resetRouteProgress
11-10 15:03:32.726: I/nav-native(28059): Ride session reset
11-10 15:03:32.727: D/MbxNavigationTelemetry(28059): sessionStop
11-10 15:03:32.727: D/MbxNavigationTelemetry(28059): flush buffer. Pending events = 0
11-10 15:03:32.727: D/MbxNavigationTelemetry(28059): resetOriginalRoute
11-10 15:03:32.727: D/MbxNavigationTelemetry(28059): resetRouteProgress
11-10 15:03:32.727: D/MbxNavigationTelemetry(28059): MapboxMetricsReporter disable
11-10 15:03:32.772: I/GCoreUlr(23206): WorldUpdater received intent Intent { act=com.google.android.location.reporting.ACTION_UPDATE_WORLD cmp=com.google.android.gms/com.google.android.location.reporting.service.DispatchingService (has extras) } with receiverAction android.location.PROVIDERS_CHANGED [CONTEXT service_id=22 ]
11-10 15:03:32.791: I/GCoreUlr(23206): WorldUpdater:android.location.PROVIDERS_CHANGED: Ensuring that reporting is stopped because of reasons: {account#-16#=[InactiveReason{mIdentifier=6, mName='ReportingNotEnabled'}, InactiveReason{mIdentifier=14, mName='HistoryNotEnabled'}]} [CONTEXT service_id=22 ]
11-10 15:03:32.808: D/LifecycleMonitor(28059): Lifecycle status change: com.mapbox.navigation.instrumentation_tests.activity.EmptyTestActivity@fb6e1b0 in: PAUSED
11-10 15:03:32.808: D/LifecycleMonitor(28059): running callback: androidx.test.rule.ActivityTestRule$LifecycleCallback@d12f54e
11-10 15:03:32.808: D/LifecycleMonitor(28059): callback completes: androidx.test.rule.ActivityTestRule$LifecycleCallback@d12f54e
11-10 15:03:32.819: E/TestRunner(28059): failed: expect_initial_alternative_route_removed_after_navigating_past(com.mapbox.navigation.instrumentation_tests.core.RouteAlternativesTest)
11-10 15:03:32.819: E/TestRunner(28059): ----- begin exception -----
11-10 15:03:32.827: E/TestRunner(28059): java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
11-10 15:03:32.827: E/TestRunner(28059): 	at androidx.test.espresso.Espresso.onIdle(Espresso.java:295)
11-10 15:03:32.827: E/TestRunner(28059): 	at androidx.test.espresso.Espresso.onIdle(Espresso.java:314)
11-10 15:03:32.827: E/TestRunner(28059): 	at com.mapbox.navigation.instrumentation_tests.core.RouteAlternativesTest$expect_initial_alternative_route_removed_after_navigating_past$3.invoke(RouteAlternativesTest.kt:104)
11-10 15:03:32.827: E/TestRunner(28059): 	at com.mapbox.navigation.instrumentation_tests.core.RouteAlternativesTest$expect_initial_alternative_route_removed_after_navigating_past$3.invoke(RouteAlternativesTest.kt:103)
11-10 15:03:32.827: E/TestRunner(28059): 	at com.mapbox.navigation.instrumentation_tests.utils.history.MapboxHistoryTestRule.stopRecordingOnCrash(MapboxHistoryTestRule.kt:59)
11-10 15:03:32.827: E/TestRunner(28059): 	at com.mapbox.navigation.instrumentation_tests.core.RouteAlternativesTest.expect_initial_alternative_route_removed_after_navigating_past(RouteAlternativesTest.kt:103)
11-10 15:03:32.827: E/TestRunner(28059): 	at java.lang.reflect.Method.invoke(Native Method)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
11-10 15:03:32.827: E/TestRunner(28059): 	at androidx.test.internal.runner.junit4.statement.RunBefores.evaluate(RunBefores.java:80)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
11-10 15:03:32.827: E/TestRunner(28059): 	at androidx.test.rule.GrantPermissionRule$RequestPermissionStatement.evaluate(GrantPermissionRule.java:134)
11-10 15:03:32.827: E/TestRunner(28059): 	at androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:531)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
11-10 15:03:32.827: E/TestRunner(28059): 	at com.schibsted.spain.barista.rule.cleardata.ClearDatabaseRule$apply$1.evaluate(ClearDatabaseRule.kt:30)
11-10 15:03:32.827: E/TestRunner(28059): 	at com.schibsted.spain.barista.rule.cleardata.ClearPreferencesRule$1.evaluate(ClearPreferencesRule.java:26)
11-10 15:03:32.827: E/TestRunner(28059): 	at com.schibsted.spain.barista.rule.cleardata.ClearFilesRule$apply$1.evaluate(ClearFilesRule.kt:25)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.Suite.runChild(Suite.java:128)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.Suite.runChild(Suite.java:27)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
11-10 15:03:32.827: E/TestRunner(28059): 	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
11-10 15:03:32.827: E/TestRunner(28059): 	at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
11-10 15:03:32.827: E/TestRunner(28059): 	at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:395)
11-10 15:03:32.827: E/TestRunner(28059): 	at android.app.Instrumentation$In
11-10 15:03:32.827: E/TestRunner(28059): ----- end exception -----
11-10 15:03:32.827: D/OrchestrationListener(28059): Sending TestFailure event: java.util.concurrent.ExecutionException: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
```

Guardiola31337 commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/LocationEx.kt:
```diff
@@ -48,11 +51,32 @@ internal fun Location.toFixLocation(): FixLocation {
+    val iterator: Iterator<String> = keySet.iterator()
+    while (iterator.hasNext()) {
+        val key = iterator.next()
+        map[key] = Value(this.getString(key) ?: "")
```
`Value` generic constructor is `private` 😬 

```java
private Value(@Nullable Object obj) {
    this.contents = obj;
}
```

Should we make it `public`?

Guardiola31337 replied:
@mapbox/core-sdk Any opinions? It's a `bindgen` type.

Guardiola31337 commented:
I can reproduce the `expect_initial_alternative_route_removed_after_navigating_past` locally

```
java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
at androidx.test.espresso.Espresso.onIdle(Espresso.java:295)
at androidx.test.espresso.Espresso.onIdle(Espresso.java:314)
at com.mapbox.navigation.instrumentation_tests.core.RouteAlternativesTest$expect_initial_alternative_route_removed_after_navigating_past$3.invoke(RouteAlternativesTest.kt:104)
at com.mapbox.navigation.instrumentation_tests.core.RouteAlternativesTest$expect_initial_alternative_route_removed_after_navigating_past$3.invoke(RouteAlternativesTest.kt:103)
at com.mapbox.navigation.instrumentation_tests.utils.history.MapboxHistoryTestRule.stopRecordingOnCrash(MapboxHistoryTestRule.kt:59)
at com.mapbox.navigation.instrumentation_tests.core.RouteAlternativesTest.expect_initial_alternative_route_removed_after_navigating_past(RouteAlternativesTest.kt:103)
at java.lang.reflect.Method.invoke(Native Method)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
at androidx.test.internal.runner.junit4.statement.RunBefores.evaluate(RunBefores.java:80)
at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
at androidx.test.rule.GrantPermissionRule$RequestPermissionStatement.evaluate(GrantPermissionRule.java:134)
at androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:531)
at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
at com.schibsted.spain.barista.rule.cleardata.ClearDatabaseRule$apply$1.evaluate(ClearDatabaseRule.kt:30)
at com.schibsted.spain.barista.rule.cleardata.ClearPreferencesRule$1.evaluate(ClearPreferencesRule.java:26)
at com.schibsted.spain.barista.rule.cleardata.ClearFilesRule$apply$1.evaluate(ClearFilesRule.kt:25)
at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
at org.junit.rules.RunRules.evaluate(RunRules.java:20)
at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:395)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:1879)
Caused by: java.util.concurrent.ExecutionException: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
at java.util.concurrent.FutureTask.report(FutureTask.java:94)
at java.util.concurrent.FutureTask.get(FutureTask.java:164)
at androidx.test.espresso.Espresso.onIdle(Espresso.java:287)
... 48 more
Caused by: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
at androidx.test.espresso.base.Interrogator.getNextMessage(Interrogator.java:205)
at androidx.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:140)
at androidx.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:519)
at androidx.test.espresso.base.UiControllerImpl.loopMainThreadUntilIdle(UiControllerImpl.java:441)
at androidx.test.espresso.Espresso$1.run(Espresso.java:277)
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:423)
at java.util.concurrent.FutureTask.run(FutureTask.java:237)
at android.os.Handler.handleCallback(Handler.java:739)
at android.os.Handler.dispatchMessage(Handler.java:95)
at android.os.Looper.loop(Looper.java:148)
at android.app.ActivityThread.main(ActivityThread.java:5417)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
Caused by: java.lang.reflect.InvocationTargetException
at java.lang.reflect.Method.invoke(Native Method)
at androidx.test.espresso.base.Interrogator.getNextMessage(Interrogator.java:199)
... 13 more
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'com.mapbox.api.directions.v5.models.DirectionsRoute$Builder com.mapbox.api.directions.v5.models.DirectionsRoute.toBuilder()' on a null object reference
at com.mapbox.api.directions.v5.models.DirectionsResponse.fromJson(DirectionsResponse.java:169)
at com.mapbox.navigation.base.internal.utils.DirectionsResponseUtilsKt$parseDirectionsResponse$2.invokeSuspend(DirectionsResponseUtils.kt:29)
at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

Under investigation.

cc @kmadsen

Guardiola31337 commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/LocationEx.kt:
```diff
@@ -48,11 +51,48 @@ internal fun Location.toFixLocation(): FixLocation {
+    val bundle = Bundle()
+    for ((key, value) in this.entries) {
+        when (val contents = value.contents) {
+            is Double -> bundle.putDouble(key, contents)
+            is Long -> bundle.putLong(key, contents)
+            is Boolean -> bundle.putBoolean(key, contents)
+            is String -> bundle.putString(key, contents)
+            else -> {
+                // do nothing
+            }
```
It'd be great if these https://github.com/mapbox/mapbox-maps-android/blob/main/extension-style/src/main/java/com/mapbox/maps/extension/style/utils/TypeUtils.kt were exposed internally so we don't have to implement them on each SDK.

cc @mr1sunshine

Guardiola31337 commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/LocationEx.kt:
```diff
@@ -48,11 +51,48 @@ internal fun Location.toFixLocation(): FixLocation {
+    val iterator: Iterator<String> = keySet.iterator()
+    while (iterator.hasNext()) {
+        val key = iterator.next()
+        when (val value = this.get(key)) {
+            is Double -> map[key] = Value(value)
+            is Long -> map[key] = Value(value)
+            is Boolean -> map[key] = Value(value)
+            is String -> map[key] = Value(value)
+            else -> {
+                // do nothing
+            }
```
Same as in https://github.com/mapbox/mapbox-navigation-android/pull/5034/files#r747684218

Guardiola31337 commented:
Fixing https://github.com/mapbox/mapbox-navigation-android/pull/5034#issuecomment-966475400 in https://github.com/mapbox/mapbox-navigation-android/pull/5034/commits/3aaba6cccb071d5027972c05ff97931adafb39a2

NN callbacks are emitted in the same thread in which the observer was created / registered, which is not the main thread for Instrumentation tests. Adding `runOnMainSync` fixes it.

Guardiola31337 commented:
BTW @DenisPryt can you think of anything that had changed from NN that may cause this regression https://github.com/mapbox/mapbox-navigation-android/pull/5034#issuecomment-966530730? Wondering if we may be workarounding the tests for now but still hiding another issue.

cc @kmadsen

Guardiola31337 commented:
OK this is ready for a final review @mapbox/navigation-android 

cc @axti @DenisPryt @SiarheiFedartsou

axti commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/LocationEx.kt:
```diff
@@ -48,11 +51,48 @@ internal fun Location.toFixLocation(): FixLocation {
         speedAccuracy,
-        verticalAccuracy
+        verticalAccuracy,
+        if (extras != null) extras.toMap() else Bundle().toMap()
```
This could be simplified like: `extras?.toMap() ?: Bundle().toMap()`

axti commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/LocationEx.kt:
```diff
@@ -48,11 +51,48 @@ internal fun Location.toFixLocation(): FixLocation {
 private fun isCurrentSdkVersionEqualOrGreaterThan(sdkCode: Int): Boolean =
     Build.VERSION.SDK_INT >= sdkCode
+
+internal fun HashMap<String, Value>.toBundle(): Bundle? {
```
It will be nice to have some alias for HashMap<String, Value> what do you think?

axti replied:
```suggestion
fun Map<String, Value>.toBundle(bundle: Bundle = Bundle()): Bundle = bundle.apply {
        forEach {
            when (val contents = it.value.contents) {
                is Double -> putDouble(it.key, contents)
                is Long -> putLong(it.key, contents)
                is Boolean -> putBoolean(it.key, contents)
                is String -> putString(it.key, contents) }
            }
        }
```

axti commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/LocationEx.kt:
```diff
@@ -48,11 +51,48 @@ internal fun Location.toFixLocation(): FixLocation {
 private fun isCurrentSdkVersionEqualOrGreaterThan(sdkCode: Int): Boolean =
     Build.VERSION.SDK_INT >= sdkCode
+
+internal fun HashMap<String, Value>.toBundle(): Bundle? {
```
I'm not sure that this is a good place to have extension functions for Bundle and HashMap inside Location.Ex.kt

Guardiola31337 replied:
Let's keep it for now, it's `internal` and as I mentioned it'd be great to have these exposed https://github.com/mapbox/mapbox-navigation-android/pull/5034#discussion_r747684218

axti commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/LocationEx.kt:
```diff
@@ -48,11 +51,48 @@ internal fun Location.toFixLocation(): FixLocation {
     Build.VERSION.SDK_INT >= sdkCode
+
+internal fun HashMap<String, Value>.toBundle(): Bundle? {
+    if (this.isEmpty()) return null
```
This looks incorrect to me. 
When we convert empty Bundle to the HashMap we get empty HashMap, but when we convert same empty HashMap back to Bundle we get null.

Guardiola31337 commented:
@axti I've just addressed the feedback. Could you take a look when you have a chance? 🙏

Guardiola31337 commented:
`expect_initial_alternative_route_removed_after_navigating_past` failed again https://console.firebase.google.com/u/1/project/mapbox-navigation-android/testlab/histories/bh.54eb2dd57e959221/matrices/8151793691623597249/executions/bs.88d7cb7dbe2c489/testcases/6 even with the `runOnMainSync` 😬 

```
11-12 07:49:09.130: D/MbxNavigationTelemetry(26191): session state is FreeDrive(sessionId=548c3dbd-37c0-4524-ae8b-17851bd53b73)
11-12 07:49:09.130: D/MbxNavigationTelemetry(26191): resetOriginalRoute
11-12 07:49:09.130: D/MbxNavigationTelemetry(26191): resetRouteProgress
11-12 07:49:09.136: D/MbxNavigationTelemetry(26191): trackFreeDrive START
11-12 07:49:09.137: D/MbxNavigationTelemetry(26191): createFreeDriveEvent START
11-12 07:49:09.163: D/MbxNavigationTelemetry(26191): populateFreeDriveEvent
11-12 07:49:09.167: D/MbxNavigationTelemetry(26191): class com.mapbox.navigation.core.telemetry.events.NavigationFreeDriveEvent event sent
11-12 07:49:09.182: D/MbxNavigationTelemetry(26191): Lifecycle monitor created
11-12 07:49:09.203: W/nav-native(26191): Location at timestamp 1537310947387ns was filtered out.
11-12 07:49:09.204: W/nav-native(26191): Got location timestamp from the past: 1538310947387ns <= 1538310947387ns. No status will be produced.
11-12 07:49:09.212: D/MbxNavigationTelemetry(26191): session state is ActiveGuidance(sessionId=295a7c4d-caba-4b09-802b-5eb3266aeea5)
11-12 07:49:09.212: D/MbxNavigationTelemetry(26191): trackFreeDrive STOP
11-12 07:49:09.212: D/MbxNavigationTelemetry(26191): createFreeDriveEvent STOP
11-12 07:49:09.240: V/NotificationService(861): pkg=com.mapbox.navigation.instrumentation_tests canInterrupt=false intercept=true
11-12 07:49:09.247: D/MbxNavigationTelemetry(26191): populateFreeDriveEvent
11-12 07:49:09.248: D/MbxNavigationTelemetry(26191): class com.mapbox.navigation.core.telemetry.events.NavigationFreeDriveEvent event sent
11-12 07:49:09.249: D/MbxNavigationTelemetry(26191): onRoutesChanged. Number of routes = 2; reason = ROUTES_UPDATE_REASON_NEW
11-12 07:49:09.249: D/MbxNavigationTelemetry(26191): handle a new route
11-12 07:49:09.249: D/MbxNavigationTelemetry(26191): resetOriginalRoute
11-12 07:49:09.249: D/MbxNavigationTelemetry(26191): resetRouteProgress
11-12 07:49:09.381: W/Looper(21155): Looper already prepared for this thread with a different value for the LOOPER_PREPARE_ALLOW_NON_CALLBACKS option.
11-12 07:49:09.491: I/art(26191): Background partial concurrent mark sweep GC freed 106223(5MB) AllocSpace objects, 38(1396KB) LOS objects, 40% free, 19MB/33MB, paused 1.954ms total 160.307ms
11-12 07:49:09.497: I/nav-native(26191): Waypoints were not passed to setRoute. Trying to restore it from route itself.
11-12 07:49:09.497: W/nav-native(26191): Cannot find via_waypoints in passed route. On-board waypoints matching will be used.
11-12 07:49:09.497: W/nav-native(26191): Cannot find via_waypoints in passed route. On-board waypoints matching will be used.
11-12 07:49:09.497: I/nav-native(26191): Successful route response. Metadata: 
11-12 07:49:09.528: D/MbxNavigationTelemetry(26191): populateNavigationEvent
11-12 07:49:09.529: D/MbxNavigationTelemetry(26191): class com.mapbox.navigation.core.telemetry.events.NavigationDepartEvent event sent
11-12 07:49:09.537: I/art(26191): Note: end time exceeds epoch: 
11-12 07:49:09.617: W/System.err(26191): java.lang.NullPointerException: Attempt to invoke virtual method 'com.mapbox.api.directions.v5.models.DirectionsRoute$Builder com.mapbox.api.directions.v5.models.DirectionsRoute.toBuilder()' on a null object reference
11-12 07:49:09.632: W/System.err(26191): 	at com.mapbox.api.directions.v5.models.DirectionsResponse.fromJson(DirectionsResponse.java:169)
11-12 07:49:09.633: W/System.err(26191): 	at com.mapbox.navigation.base.internal.utils.DirectionsResponseUtilsKt$parseDirectionsResponse$2.invokeSuspend(DirectionsResponseUtils.kt:29)
11-12 07:49:09.633: W/System.err(26191): 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
11-12 07:49:09.633: W/System.err(26191): 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
11-12 07:49:09.633: W/System.err(26191): 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
11-12 07:49:09.633: W/System.err(26191): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
11-12 07:49:09.633: W/System.err(26191): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
11-12 07:49:09.633: W/System.err(26191): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
11-12 07:49:09.633: E/nav-native(26191): RouteAlternativesControllerImpl unhandled observer unknown exception in onRouteAlternativesChanged method
11-12 07:49:09.655: W/LocationUpdateReceiver(26191): LocationEngineResult == null
11-12 07:49:09.657: V/NotificationService(861): pkg=com.mapbox.navigation.instrumentation_tests canInterrupt=false intercept=true
11-12 07:49:09.658: E/DEBUG(26191): alternatives failed history path=/data/user/0/com.mapbox.navigation.instrumentation_tests/files/mbx_nav/history/2021-11-12T15-49-09Z_e9a986b8-1892-4f36-88c5-07dc9f5c99cf.pbf.gz
11-12 07:49:09.663: I/MbxHistoryRecorder(26191): Initial size is 4096 bytes
11-12 07:49:09.669: D/MbxNavigationTelemetry(26191): MapboxNavigation onDestroy
11-12 07:49:09.674: D/MbxNavigationTelemetry(26191): session state is com.mapbox.navigation.core.trip.session.NavigationSessionState$Idle@cd3f883
11-12 07:49:09.675: D/MbxNavigationTelemetry(26191): handleSessionCanceled
11-12 07:49:09.675: D/MbxNavigationTelemetry(26191): flush buffer. Pending events = 0
11-12 07:49:09.692: D/MbxNavigationTelemetry(26191): populateNavigationEvent
11-12 07:49:09.692: D/MbxNavigationTelemetry(26191): class com.mapbox.navigation.core.telemetry.events.NavigationCancelEvent event sent
11-12 07:49:09.693: D/MbxNavigationTelemetry(26191): resetOriginalRoute
11-12 07:49:09.693: D/MbxNavigationTelemetry(26191): resetRouteProgress
11-12 07:49:09.695: I/nav-native(26191): Ride session reset
11-12 07:49:09.695: D/MbxNavigationTelemetry(26191): sessionStop
11-12 07:49:09.695: D/MbxNavigationTelemetry(26191): flush buffer. Pending events = 0
11-12 07:49:09.695: D/MbxNavigationTelemetry(26191): resetOriginalRoute
11-12 07:49:09.695: D/MbxNavigationTelemetry(26191): resetRouteProgress
11-12 07:49:09.695: D/MbxNavigationTelemetry(26191): MapboxMetricsReporter disable
11-12 07:49:09.863: D/LifecycleMonitor(26191): Lifecycle status change: com.mapbox.navigation.instrumentation_tests.activity.EmptyTestActivity@d17153b in: PAUSED
11-12 07:49:09.863: D/LifecycleMonitor(26191): running callback: androidx.test.rule.ActivityTestRule$LifecycleCallback@3424e0d
11-12 07:49:09.863: D/LifecycleMonitor(26191): callback completes: androidx.test.rule.ActivityTestRule$LifecycleCallback@3424e0d
11-12 07:49:09.888: I/BackdropActivity(20666): onResume()
11-12 07:49:09.896: V/NotificationService(861): pkg=com.google.android.apps.mtaas.backdrop canInterrupt=false intercept=true
11-12 07:49:09.899: E/TestRunner(26191): failed: expect_initial_alternative_route_removed_after_navigating_past(com.mapbox.navigation.instrumentation_tests.core.RouteAlternativesTest)
```

It seems there's a race condition / issue somewhere https://github.com/mapbox/mapbox-navigation-android/pull/5034#issuecomment-966539745

@DenisPryt could you help here when you have a chance? 🙏 

cc @kmadsen

axti reviewed (approved):
LGTM

DenisPryt commented:
About expect_initial_alternative_route_removed_after_navigating_past failings:
Since version 80.0.0 routeResponse that provides from onRouteAlternativesChanged has changed. Previously it was original route request with single route. Original: `{"routes": [{r1}, {r2}], ...}`, provided: `{"routes": [{r2}], ...}`
Now we provide original route response with nulls instead of all routes except one. 
Original: `{"routes": [{r1}, {r2}], ...}`, provided: `{"routes": [null, {r2}], ...}`
This is  need to use route response from alternatives with RouterInterface.getRouteRefresh

LukasPaczos commented:
I tried doing something like this in the `parseDirectionsResponse` (which now has to be specific for alternatives):
```kotlin
val jsonRoutes = jsonObject.getJSONArray("routes")
val jsonRoute: DirectionsRoute = jsonRoutes.let {
    for (i in 0 until it.length()) {
        if (!it.isNull(i)) {
            return@let DirectionsRoute.fromJson(
                it.getJSONObject(i).toString(),
                options,
                uuid
            ).apply {
                toBuilder()
                    .routeIndex(i)
                    .build()
            }
        }
    }
    return@withContext null
}
return@withContext jsonRoute
```

Unfortunately in order to be able to support this, we'll need to expose [`DirectionsRoute.Builder#routeIndex`](https://github.com/mapbox/mapbox-java/blob/210d48b04c94500a3a9f127fc00b75956ec65704/services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/DirectionsRoute.java#L326) as it's currently package private.

Alternatively, we'd have to allow for nullable routes in the JSON which `DirectionsResponse#fromJson` currently does not support.

DenisPryt commented:
@LukasPaczos to avoid this parsing I can provide full route response without any changes. But to find correct route you need to use routeIndex anyway.

LukasPaczos commented:
That would be another option - full route response with the index of the route that we should use. Problem here might be a lot of redundant data passed and stored in memory and then a lot of redundant parsing of route objects out of which we always use only one. Having this in mind, a solution where we pass and store only one route sounds better to me.

kmadsen reviewed (approved):
Nice work on the route alternatives quick-fix!

RingerJK commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/navigator/LocationEx.kt:
```diff
@@ -4,10 +4,14 @@ package com.mapbox.navigation.core.navigator
 import com.mapbox.navigator.FixLocation
 import java.util.Date
 
+internal typealias FixLocationExtras = HashMap<String, Value>
```
value class candidate 💯
